java-util/src/main/java/io/druid/java/util/common/granularity/Granularity.java

```diff
@@ -0,0 +1,538 @@
+      if (!Objects.equals(result, toMerge.get(i))) {
+        return null;
+      }
+    }
+
+    return result;
+  }
+
+  public static List<Granularity> granularitiesFinerThan(final Granularity gran0)
+  {
+    final DateTime epoch = new DateTime(0);
+    final List<Granularity> retVal = Lists.newArrayList();
+    final DateTime origin = (gran0 instanceof PeriodGranularity) ? ((PeriodGranularity) gran0).getOrigin() : null;
+    final DateTimeZone tz = (gran0 instanceof PeriodGranularity) ? ((PeriodGranularity) gran0).getTimeZone() : null;
+    for (GranularityType gran : GranularityType.values()) {
+      if (gran == GranularityType.ALL || gran == GranularityType.NONE) {
+        continue;
+      }
+      final Granularity segmentGranularity = GranularityType.createGranularity(gran.name(), origin, tz);
+      if (segmentGranularity.bucket(epoch).toDurationMillis() <= gran0.bucket(epoch).toDurationMillis()) {
+        retVal.add(segmentGranularity);
+      }
+    }
+    Collections.sort(
+        retVal,
+        new Comparator<Granularity>()
+        {
+          @Override
+          public int compare(Granularity g1, Granularity g2)
+          {
+            return Longs.compare(g2.bucket(epoch).toDurationMillis(), g1.bucket(epoch).toDurationMillis());
+          }
+        }
+    );
+    return retVal;
+  }
+
+  public abstract DateTimeFormatter getFormatter(Formatter type);
+
+  public abstract DateTime increment(DateTime time);
+
+  public abstract DateTime decrement(DateTime time);
+
+  public abstract DateTime truncate(DateTime time);
+
+  public abstract DateTime toDate(String filePath, Formatter formatter);
+
+  public abstract byte[] cacheKey();
+
+  public DateTime toDateTime(long offset)
+  {
+    return new DateTime(offset, DateTimeZone.UTC);
+  }
+
+  public DateTime toDate(String filePath)
+  {
+    return toDate(filePath, Formatter.DEFAULT);
+  }
+
+  public final String toPath(DateTime time)
+  {
+    return toPath(time, "default");
+  }
+
+  private final String toPath(DateTime time, String type)
+  {
+    return toPath(time, Formatter.valueOf(type.toUpperCase()));
+  }
+
+  private final String toPath(DateTime time, Formatter type)
+  {
+    return getFormatter(type).print(time);
+  }
+
+  /**
+   * Return a granularity-sized Interval containing a particular DateTime.
+   */
+  public final Interval bucket(DateTime t)
+  {
+    DateTime start = truncate(t);
+    return new Interval(start, increment(start));
+  }
+
+  // Used by the toDate implementations.
+  final Integer[] getDateValues(String filePath, Formatter formatter)
+  {
+    Pattern pattern = defaultPathPattern;
+    switch (formatter) {
+      case DEFAULT:
+      case LOWER_DEFAULT:
+        break;
+      case HIVE:
+        pattern = hivePathPattern;
+        break;
+      default:
+        throw new IAE("Format %s not supported", formatter);
+    }
+
+    Matcher matcher = pattern.matcher(filePath);
+
+    Integer[] vals = new Integer[7];
```

Comment:
7 because  this array contains standard datetime field values namely:
year, monthOfYear, dayOfMonth, hourOfDay, minuteOfHour, secondOfMinute and  millisOfSecond